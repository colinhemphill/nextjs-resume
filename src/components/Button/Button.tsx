import { VariantProps, cva } from 'class-variance-authority';
import { PropsWithChildren } from 'react';
import { twMerge } from 'tailwind-merge';

export const buttonVariants = cva(
  'flex items-center justify-center gap-2 rounded-md border-2 border-accent-7 bg-accent-1 font-medium text-accent-11 no-underline outline-none transition hover:border-accent-8 focus:border-transparent focus:bg-accent-9 focus:text-accentContrast focus:ring-4 focus:ring-accent-6 hover:focus:border-transparent',
  {
    variants: {
      size: {
        sm: 'text-sm py-1 px-4',
        md: 'text-base py-2 px-6',
        lg: 'text-lg py-3 px-8',
      },
    },
    defaultVariants: {
      size: 'md',
    },
  },
);

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof buttonVariants>;

export default function Button({
  children,
  className,
  size,
  ...buttonProps
}: PropsWithChildren<ButtonProps>) {
  return (
    <button
      className={twMerge(buttonVariants({ size }), className)}
      type="button"
      {...buttonProps}
    >
      {children}
    </button>
  );
}
