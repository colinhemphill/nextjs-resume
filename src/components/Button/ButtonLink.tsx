import { VariantProps } from 'class-variance-authority';
import Link, { LinkProps } from 'next/link';
import { PropsWithChildren } from 'react';
import { twMerge } from 'tailwind-merge';
import { buttonVariants } from './Button';

export type ButtonLinkProps = LinkProps &
  VariantProps<typeof buttonVariants> & { className?: string };

export default function ButtonLink({
  children,
  className,
  size,
  ...linkProps
}: PropsWithChildren<ButtonLinkProps>) {
  return (
    <Link
      className={twMerge(buttonVariants({ size }), className)}
      {...linkProps}
    >
      {children}
    </Link>
  );
}
