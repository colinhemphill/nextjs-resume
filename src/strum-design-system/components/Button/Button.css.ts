import { style } from '@vanilla-extract/css';
import { recipe, RecipeVariants } from '@vanilla-extract/recipes';
import { atoms } from '../../sprinkles.css';
import { rounded } from '../../styles/borderRadius.css';
import { vars } from '../../themes/contract.css';

const buttonBaseStyle = style([
  rounded,
  atoms({
    paddingX: 4,
    paddingY: 2,
  }),
  {
    border: '1px solid transparent',
    cursor: 'pointer',
    display: 'inline-block',
    fontSize: '1rem',
    fontWeight: 400,
    lineHeight: 1.5,
    textAlign: 'center',
    textDecoration: 'none',
    userSelect: 'none',
    verticalAlign: 'middle',
  },
  {
    selectors: {
      '&:disabled': {
        cursor: 'default',
        opacity: 0.7,
      },
    },
  },
]);

export const buttonStyle = recipe({
  base: buttonBaseStyle,
  defaultVariants: { color: 'primary', size: 'md' },
  variants: {
    color: {
      black: {
        backgroundColor: vars.colors.black,
        borderColor: vars.colors.black,
        color: vars.contrastColors.black,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.black,
            borderColor: vars.darkenedColors.black,
          },
        },
      },
      dark: {
        backgroundColor: vars.colors.dark,
        borderColor: vars.colors.dark,
        color: vars.contrastColors.dark,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.dark,
            borderColor: vars.darkenedColors.dark,
          },
        },
      },
      light: {
        backgroundColor: vars.colors.light,
        borderColor: vars.colors.light,
        color: vars.contrastColors.light,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.light,
            borderColor: vars.darkenedColors.light,
          },
        },
      },
      medium: {
        backgroundColor: vars.colors.medium,
        borderColor: vars.colors.medium,
        color: vars.contrastColors.medium,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.medium,
            borderColor: vars.darkenedColors.medium,
          },
        },
      },
      primary: {
        backgroundColor: vars.colors.primary,
        borderColor: vars.colors.primary,
        color: vars.contrastColors.primary,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.primary,
            borderColor: vars.darkenedColors.primary,
          },
        },
      },
      white: {
        backgroundColor: vars.colors.white,
        borderColor: vars.colors.white,
        color: vars.contrastColors.white,
        selectors: {
          '&:hover:not(:disabled), &:focus:not(:disabled)': {
            backgroundColor: vars.darkenedColors.white,
            borderColor: vars.darkenedColors.white,
          },
        },
      },
    },
    size: {
      lg: { fontSize: '1.25rem' },
      md: {
        fontSize: '1rem',
      },
      sm: {
        fontSize: '0.875rem',
        padding: `${vars.spacers[1]} ${vars.spacers[2]}`,
      },
      xl: {
        fontSize: '1.5rem',
        fontWeight: 700,
        padding: `${vars.spacers[3]} ${vars.spacers[5]}`,
      },
    },
  },
});

export type ButtonVariants = RecipeVariants<typeof buttonStyle>;
