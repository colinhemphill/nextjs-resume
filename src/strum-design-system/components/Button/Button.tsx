import React, { ButtonHTMLAttributes } from 'react';
import { Atoms } from '../../sprinkles.css';
import { composeWithAtoms } from '../../utils/compose';
import { buttonStyle, ButtonVariants } from './Button.css';

interface Button extends ButtonHTMLAttributes<HTMLButtonElement> {
  atoms?: Atoms;
  color?: ButtonVariants['color'];
  size?: ButtonVariants['size'];
}

const Button: React.FC<Button> = (props) => {
  const { atoms: atomicProperties, children, color, size, ...rest } = props;

  const classes = composeWithAtoms(
    atomicProperties,
    buttonStyle({
      color,
      size,
    }),
  );

  return (
    <button {...rest} className={classes}>
      {children}
    </button>
  );
};

export default Button;
