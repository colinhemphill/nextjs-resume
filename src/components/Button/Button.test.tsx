import { render, screen } from '@test-utils';
import { describe, expect, test } from 'vitest';
import Button from './Button';

describe('<Button />', () => {
  test('Snapshot', async () => {
    const { container } = render(
      <>
        <Button>Button</Button>
        <Button size="sm">Button</Button>
        <Button size="md">Button</Button>
        <Button size="lg">Button</Button>
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Renders an accessible button element', async () => {
    render(<Button>Action</Button>);

    const button = screen.getByRole('button', { name: /action/i });
    expect(button).toBeDefined();
  });
});
