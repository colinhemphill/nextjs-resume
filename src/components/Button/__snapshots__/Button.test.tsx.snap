// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button /> > Snapshot 1`] = `
<div>
  <button
    class="flex items-center justify-center gap-2 rounded-md border-2 border-accent-7 bg-accent-1 font-medium text-accent-11 no-underline outline-none transition hover:border-accent-8 focus:border-transparent focus:bg-accent-9 focus:text-accentContrast focus:ring-4 focus:ring-accent-6 hover:focus:border-transparent text-base py-2 px-6"
    type="button"
  >
    Button
  </button>
  <button
    class="flex items-center justify-center gap-2 rounded-md border-2 border-accent-7 bg-accent-1 font-medium text-accent-11 no-underline outline-none transition hover:border-accent-8 focus:border-transparent focus:bg-accent-9 focus:text-accentContrast focus:ring-4 focus:ring-accent-6 hover:focus:border-transparent text-sm py-1 px-4"
    type="button"
  >
    Button
  </button>
  <button
    class="flex items-center justify-center gap-2 rounded-md border-2 border-accent-7 bg-accent-1 font-medium text-accent-11 no-underline outline-none transition hover:border-accent-8 focus:border-transparent focus:bg-accent-9 focus:text-accentContrast focus:ring-4 focus:ring-accent-6 hover:focus:border-transparent text-base py-2 px-6"
    type="button"
  >
    Button
  </button>
  <button
    class="flex items-center justify-center gap-2 rounded-md border-2 border-accent-7 bg-accent-1 font-medium text-accent-11 no-underline outline-none transition hover:border-accent-8 focus:border-transparent focus:bg-accent-9 focus:text-accentContrast focus:ring-4 focus:ring-accent-6 hover:focus:border-transparent text-lg py-3 px-8"
    type="button"
  >
    Button
  </button>
</div>
`;
